x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/CentroidCalculator.java

```diff
@@ -27,106 +26,64 @@
-        geometry.visit(visitor);
-        this.dimensionalShapeType = visitor.calculator.dimensionalShapeType;
-    }
 
-    /**
-     * adds a single coordinate to the running sum and count of coordinates
-     * for centroid calculation
-     *  @param x the x-coordinate of the point
-     * @param y the y-coordinate of the point
-     * @param weight the associated weight of the coordinate
-     */
-    private void addCoordinate(double x, double y, double weight, DimensionalShapeType dimensionalShapeType) {
-        // x and y can be infinite due to really small areas and rounding problems
-        if (Double.isFinite(x) && Double.isFinite(y)) {
-            if (this.dimensionalShapeType == null || this.dimensionalShapeType == dimensionalShapeType) {
-                compSumX.add(x * weight);
-                compSumY.add(y * weight);
-                compSumWeight.add(weight);
-                this.dimensionalShapeType = dimensionalShapeType;
-            } else if (dimensionalShapeType.compareTo(this.dimensionalShapeType) > 0) {
-                // reset counters
-                compSumX.reset(x * weight, 0);
-                compSumY.reset(y * weight, 0);
-                compSumWeight.reset(weight, 0);
-                this.dimensionalShapeType = dimensionalShapeType;
-            }
-        }
+    private final CentroidCalculatorVisitor visitor;
+
+    public CentroidCalculator() {
+        this.visitor = new CentroidCalculatorVisitor();
     }
 
     /**
-     * Adjusts the existing calculator to add the running sum and count
-     * from another {@link CentroidCalculator}. This is used to keep
-     * a running count of points from different sub-shapes of a single
-     * geo-shape field
+     * Add a geometry to the calculator
      *
-     * @param otherCalculator the other centroid calculator to add from
+     * @param geometry the geometry to add
      */
-    public void addFrom(CentroidCalculator otherCalculator) {
-        int compared = dimensionalShapeType.compareTo(otherCalculator.dimensionalShapeType);
-        if (compared < 0) {
-            dimensionalShapeType = otherCalculator.dimensionalShapeType;
-            this.compSumX = otherCalculator.compSumX;
-            this.compSumY = otherCalculator.compSumY;
-            this.compSumWeight = otherCalculator.compSumWeight;
-
-        } else if (compared == 0) {
-            this.compSumX.add(otherCalculator.compSumX.value());
-            this.compSumY.add(otherCalculator.compSumY.value());
-            this.compSumWeight.add(otherCalculator.compSumWeight.value());
-        } // else (compared > 0) do not modify centroid calculation since otherCalculator is of lower dimension than this calculator
+    public void add(Geometry geometry) {
+        geometry.visit(visitor);
     }
 
     /**
      * @return the x-coordinate centroid
      */
     public double getX() {
         // normalization required due to floating point precision errors
```

Comment:
I remember doing this because sometimes the division would lead to degrees right above 180 or 90. Is this not needed anymore? Should we remove this since it is no longer the concern of the calculator to normalize, and instead, the encoder will be doing this?